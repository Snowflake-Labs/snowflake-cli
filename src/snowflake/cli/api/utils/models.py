import os


class DictWithEnvironFallback(dict):
    def __getattr__(self, item):
        return self[item]

    def __getitem__(self, item):
        if item in os.environ:
            return os.environ[item]
        return super().__getitem__(item)

    def __contains__(self, item):
        return item in os.environ or super().__contains__(item)
