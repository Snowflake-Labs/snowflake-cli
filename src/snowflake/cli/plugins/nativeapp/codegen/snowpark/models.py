from __future__ import annotations

from enum import Enum
from typing import List, Optional

from pydantic import Field
from snowflake.cli.api.project.schemas.snowpark.callable import _CallableBase
from snowflake.cli.api.project.schemas.updatable_model import IdentifierField


class ExtensionFunctionTypeEnum(str, Enum):
    PROCEDURE = "PROCEDURE"
    FUNCTION = "FUNCTION"
    TABLE_FUNCTION = "TABLE_FUNCTION"
    AGGREGATE_FUNCTION = "AGGREGATE_FUNCTION"


class NativeAppExtensionFunction(_CallableBase):
    function_type: ExtensionFunctionTypeEnum
    lineno: Optional[int] = Field(
        title="The line number of the extension function", default=None
    )
    name: Optional[str] = Field(title="The name of the external function", default=None)
    packages: Optional[List[str]] = Field(
        title="List of packages (with optional version constraints) to be loaded for the function",
        default={},
    )
    schema_name: Optional[str] = IdentifierField(
        title=f"Name of the schema for the function",
        default=None,
        alias="schema",
    )
    application_roles: Optional[List[str]] = Field(
        title="Application roles granted usage to the function",
        default=[],
    )
    execute_as_caller: Optional[bool] = Field(
        title="Determine whether the procedure is executed with the privileges of "
        "the owner or with the privileges of the caller",
        default=None,
    )
