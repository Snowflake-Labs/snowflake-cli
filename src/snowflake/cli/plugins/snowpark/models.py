from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import List

from requirements import requirement


class PypiOption(Enum):
    YES = "yes"
    NO = "no"
    ASK = "ask"


class Requirement(requirement.Requirement):
    @classmethod
    def parse_line(cls, line: str) -> Requirement:
        result = super().parse_line(line)

        if result.uri and not result.name:
            result.name = result.uri

        return result


@dataclass
class SplitRequirements:
    """A dataclass to hold the results of parsing requirements files and dividing them into
    snowflake-supported vs other packages.
    """

    snowflake: List[Requirement]
    other: List[Requirement]


@dataclass
class RequirementWithFiles:
    """A dataclass to hold a requirement and the path to the
    downloaded files/folders that belong to it"""

    requirement: Requirement
    files: List[str]


@dataclass
class RequirementWithFilesAndDeps(RequirementWithFiles):
    dependencies: List[str]

    def to_requirement_with_files(self):
        return RequirementWithFiles(self.requirement, self.files)


pip_failed_msg = """pip failed with return code {}.
            If pip is installed correctly, this may mean you`re trying to install a package
            that isn't compatible with the host architecture -
            and generally means it has native libraries."""
second_chance_msg = """Good news! The following package dependencies can be
                imported directly from Anaconda, and will be excluded from
                the zip: {}"""
