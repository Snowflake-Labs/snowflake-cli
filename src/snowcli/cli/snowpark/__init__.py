import typer

from snowcli.cli.snowpark.function import app as function_app
from snowcli.cli.snowpark.package import app as package_app
from snowcli.cli.snowpark.procedure import app as procedure_app

app = typer.Typer(
    name="snowpark",
    context_settings={"help_option_names": ["-h", "--help"]},
    help="Manage functions, procedures and Snowpark objects",
)

app.add_typer(function_app)
app.add_typer(package_app)
app.add_typer(procedure_app)
