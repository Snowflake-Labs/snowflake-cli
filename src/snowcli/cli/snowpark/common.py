from __future__ import annotations

import os
from typing import Dict, List, Optional

from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowcli.cli.constants import ObjectNames, ObjectType
from snowcli.utils import generate_deploy_stage_name
from snowflake.connector.cursor import DictCursor, SnowflakeCursor


def _map_type(type_: str):
    mapping = {"string": "varchar"}
    return mapping.get(type_.lower(), type_)


def object_to_signature(function_or_procedure: Dict):
    signature = ", ".join(
        _map_type(o["type"]) for o in function_or_procedure["signature"]
    )
    return f"{function_or_procedure['name']}({signature}) RETURN {_map_type(function_or_procedure['returns'])}".upper()


def remove_parameter_names(identifier: str):
    """
    Removes parameter names from identifier.
    Deploy commands for function and procedure requires identifier
    with parameter names (e.g. `hello(number int, name string)`),
    but describe command requires only parameter types (e.g. `hello(int, string)`)
    :param identifier: `hello(number int, name string)`
    :return: `hello(int, string)`
    """
    open_parenthesis_index = identifier.index("(")
    parameters = identifier[open_parenthesis_index + 1 : -1]
    if not parameters:
        return identifier
    types = [t.strip().split(" ")[1] for t in parameters.split(",")]
    return f"{identifier[0:open_parenthesis_index]}({', '.join(types)})"


def check_if_replace_is_required(
    object_type: ObjectType,
    current_state,
    handler: str,
    return_type: str,
) -> bool:
    import logging

    log = logging.getLogger(__name__)
    resource_json = _convert_resource_details_to_dict(current_state)
    anaconda_packages = resource_json["packages"]
    log.info(
        f"Found {len(anaconda_packages)} defined Anaconda "
        f"packages in deployed {object_type}..."
    )
    log.info("Checking if app configuration has changed...")
    updated_package_list = _get_snowflake_packages_delta(
        anaconda_packages,
    )

    if updated_package_list:
        diff = len(updated_package_list) - len(anaconda_packages)
        log.info(f"Found difference of {diff} packages. Replacing the {object_type}.")
        return True

    if (
        resource_json["handler"].lower() != handler.lower()
        or _sql_to_python_return_type_mapper(resource_json["returns"]).lower()
        != return_type.lower()
    ):
        log.info(
            f"Return type or handler types do not match. Replacing the {object_type}."
        )
        return True

    return False


def _convert_resource_details_to_dict(function_details: SnowflakeCursor) -> dict:
    import json

    function_dict = {}
    json_properties = ["packages", "installed_packages"]
    for function in function_details:
        if function[0] in json_properties:
            function_dict[function[0]] = json.loads(
                function[1].replace("'", '"'),
            )
        else:
            function_dict[function[0]] = function[1]
    return function_dict


def _get_snowflake_packages_delta(anaconda_packages) -> List[str]:
    updated_package_list = []
    if os.path.exists("requirements.snowflake.txt"):
        with open("requirements.snowflake.txt", encoding="utf-8") as f:
            # for each line, check if it exists in anaconda_packages. If it
            # doesn't, add it to the return string
            for line in f:
                if line.strip() not in anaconda_packages:
                    updated_package_list.append(line.strip())
        return updated_package_list
    else:
        return updated_package_list


def _sql_to_python_return_type_mapper(resource_return_type: str) -> str:
    """
    Some of the python data types get converted to SQL types, when function/procedure is created.
    So, to properly compare types, we use mapping based on:
    https://docs.snowflake.com/en/developer-guide/udf-stored-procedure-data-type-mapping#sql-python-data-type-mappings

    Mind you, this only applies to cases, in which Snowflake accepts python type as return.
    Ie. if function returns list, it has to be declared as 'array' during creation,
    therefore any conversion is not necessary
    """
    mapping = {
        "number(38,0)": "int",
        "timestamp_ntz(9)": "datetime",
        "timestamp_tz(9)": "datetime",
        "varchar(16777216)": "string",
    }

    return mapping.get(resource_return_type.lower(), resource_return_type.lower())


class SnowparkObjectManager(SqlExecutionMixin):
    @property
    def _object_type(self) -> ObjectType:
        raise NotImplementedError()

    @property
    def _object_execute(self):
        raise NotImplementedError()

    def create(self, *args, **kwargs) -> SnowflakeCursor:
        raise NotImplementedError()

    def execute(self, execution_identifier: str) -> SnowflakeCursor:
        return self._execute_query(f"{self._object_execute} {execution_identifier}")

    @staticmethod
    def artifact_stage_path(identifier: str):
        return generate_deploy_stage_name(identifier).lower()

    def get_existing_objects(self):
        """
        Returns list of existing objects identified by their signature, for example ['HELLO(VARCHAR) RETURN VARCHAR']
        """
        results = self._execute_query(
            f"show user {self._object_type.value.sf_plural_name}",
            cursor_class=DictCursor,
        )
        return {o["arguments"]: o for o in results}

    def create_query(
        self,
        identifier: str,
        return_type: str,
        handler: str,
        artifact_file: str,
        packages: List[str],
        external_access_integrations: Optional[List[str]] = None,
        secrets: Optional[Dict[str, str]] = None,
        execute_as_caller: bool = False,
    ) -> str:
        packages_list = ",".join(f"'{p}'" for p in packages)
        query = [
            f"create or replace {self._object_type.value.sf_name} {identifier}",
            f"returns {return_type}",
            "language python",
            "runtime_version=3.8",
            f"imports=('{artifact_file}')",
            f"handler='{handler}'",
            f"packages=({packages_list})",
        ]

        if external_access_integrations:
            external_access_integration_name = ",".join(
                f"{e}" for e in external_access_integrations
            )
            query.append(
                f"external_access_integrations=({external_access_integration_name})"
            )

        if secrets:
            secret_name = ",".join(f"'{k}'={v}" for k, v in secrets.items())
            query.append(f"secrets=({secret_name})")

        if execute_as_caller:
            query.append("execute as caller")

        return "\n".join(query)


def build_udf_sproc_identifier(udf_sproc_dict):
    arguments = ", ".join(
        (f"{arg['name']} {arg['type']}" for arg in udf_sproc_dict["signature"])
    )
    return f"{udf_sproc_dict['name']}({arguments})"
