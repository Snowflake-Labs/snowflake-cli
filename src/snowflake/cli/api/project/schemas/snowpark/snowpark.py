from __future__ import annotations

from typing import List

from pydantic import Field
from snowflake.cli.api.project.schemas.snowpark.callable import (
    FunctionSchema,
    ProcedureSchema,
)
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel


class Snowpark(UpdatableModel):
    project_name: str = Field(title="Project identifier")
    stage_name: str = Field(title="Stage in which projectâ€™s artifacts will be stored")
    src: str = Field(title="Folder where your code should be located")
    functions: List[FunctionSchema] | None = Field(
        title="List of functions defined in the project", default=[]
    )
    procedures: List[ProcedureSchema] | None = Field(
        title="List of procedures defined in the project", default=[]
    )
