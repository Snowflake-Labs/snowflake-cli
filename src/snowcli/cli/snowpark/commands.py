from __future__ import annotations

import logging
import os
from enum import Enum
from pathlib import Path
from tempfile import TemporaryDirectory
from typing import Dict, List, Set

import snowcli.utils.file_utils
import typer
from click import ClickException
from snowcli.cli.common.cli_global_context import cli_context
from snowcli.cli.common.decorators import (
    global_options,
    global_options_with_connection,
    with_project_definition,
)
from snowcli.cli.common.flags import (
    DEFAULT_CONTEXT_SETTINGS,
    execution_identifier_argument,
)
from snowcli.cli.common.project_initialisation import add_init_command
from snowcli.cli.constants import DEPLOYMENT_STAGE, ObjectType
from snowcli.cli.object.manager import ObjectManager
from snowcli.cli.object.stage.manager import StageManager
from snowcli.cli.snowpark.common import (
    build_udf_sproc_identifier,
    check_if_replace_is_required,
    remove_parameter_names,
)
from snowcli.cli.snowpark.manager import FunctionManager, ProcedureManager
from snowcli.cli.snowpark_shared import (
    CheckAnacondaForPyPiDependencies,
    PackageNativeLibrariesOption,
    PyPiDownloadOption,
    snowpark_package,
)
from snowcli.exception import (
    SecretsWithoutExternalAccessIntegrationError,
)
from snowcli.output.decorators import with_output
from snowcli.output.types import (
    CollectionResult,
    CommandResult,
    MessageResult,
    SingleQueryResult,
)
from snowcli.utils import utils
from snowcli.utils.package_utils import get_snowflake_packages
from snowcli.utils.zipper import add_file_to_existing_zip
from snowflake.connector import DictCursor, ProgrammingError

log = logging.getLogger(__name__)

app = typer.Typer(
    name="snowpark",
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    help="Manage procedures and functions.",
)

ReplaceOption = typer.Option(
    False,
    "--replace",
    help="Replace procedure or function, even if no detected changes to metadata",
)

InstallCoverageWrapper = typer.Option(
    False,
    "--install-coverage-wrapper",
    help="Whether to wrap the procedure with a code coverage measurement tool, "
    "so a coverage report can be later retrieved.",
)

ObjectTypeArgument = typer.Argument(
    help="Type of snowpark object",
    case_sensitive=False,
)

add_init_command(app, project_type="snowpark", template="default_snowpark")


def _alter_procedure_artifact_with_coverage_wrapper(
    artifact_stage_path: str,
    artifact_path: Path,
    handler: str,
    identifier: str,
    stage_name: str,
):
    signature_start_index = identifier.index("(")
    name = identifier[0:signature_start_index]
    signature = identifier[signature_start_index:]
    stage_directory = artifact_stage_path.rpartition("/")[0]
    handler = _replace_handler_in_zip(
        proc_name=name,
        proc_signature=signature,
        handler=handler,
        coverage_reports_stage=stage_name,
        coverage_reports_stage_path=f"{stage_directory}/coverage",
        zip_file_path=str(artifact_path),
    )

    return handler


@app.command("deploy")
@with_output
@with_project_definition("snowpark")
@global_options_with_connection
def deploy(
    install_coverage_wrapper: bool = InstallCoverageWrapper,
    replace: bool = ReplaceOption,
    **options,
) -> CommandResult:
    """
    Deploys procedures and functions defined in project. Deploying the project alters all object defined in it.
    By default, if any of the objects exist already the commands will fail unless `--replace` flag is provided.
    All deployed object use the same artefact which is deployed only once.
    """
    snowpark = cli_context.project_definition

    procedures = snowpark.get("procedures", [])
    functions = snowpark.get("functions", [])

    if not procedures and not functions:
        raise ClickException(
            "No procedures or functions were specified in project definition."
        )

    if (len(procedures) > 1 or functions) and install_coverage_wrapper:
        raise ClickException(
            "Using coverage wrapper is currently limited to project with single procedure."
        )

    build_artifact_path = _get_snowpark_artefact_path(snowpark)

    if not build_artifact_path.exists():
        raise ClickException(
            "Artifact required for deploying the project does not exist in this directory. "
            "Please use build command to create it."
        )

    pm = ProcedureManager()
    fm = FunctionManager()
    om = ObjectManager()

    _check_if_all_defined_integrations_exists(om, functions, procedures)

    existing_functions = _find_existing_objects(ObjectType.FUNCTION, functions, om)
    existing_procedures = _find_existing_objects(ObjectType.PROCEDURE, procedures, om)

    if (existing_functions or existing_procedures) and not replace:
        msg = "Following objects already exists. Consider using --replace.\n"
        msg += "\n".join(f"function: {n}" for n in existing_functions)
        msg += "\n".join(f"procedure: {n}" for n in existing_procedures)
        raise ClickException(msg)

    # Create stage
    stage_name = snowpark.get("stage_name", DEPLOYMENT_STAGE)
    stage_manager = StageManager()
    stage_manager.create(
        stage_name=stage_name, comment="deployments managed by snowcli"
    )

    packages = get_snowflake_packages()

    artifact_stage_directory = get_app_stage_path(snowpark)
    artifact_stage_target = f"{artifact_stage_directory}/{build_artifact_path.name}"

    # Coverage case
    if install_coverage_wrapper:
        return _deploy_procedure_with_coverage(
            artifact_stage_directory=artifact_stage_directory,
            artifact_stage_target=artifact_stage_target,
            build_artifact_path=build_artifact_path,
            packages=packages,
            pm=pm,
            procedure=procedures[0],
            stage_manager=stage_manager,
            stage_name=stage_name,
            existing_objects=existing_procedures,
        )

    stage_manager.put(
        local_path=build_artifact_path,
        stage_path=artifact_stage_directory,
        overwrite=True,
    )

    deploy_status = []
    # Procedures
    for procedure in procedures:
        operation_result = _deploy_single_object(
            manager=pm,
            object_type=ObjectType.PROCEDURE,
            object_definition=procedure,
            existing_objects=existing_procedures,
            packages=packages,
            stage_artifact_path=artifact_stage_target,
        )
        deploy_status.append(operation_result)

    # Functions
    for function in functions:
        operation_result = _deploy_single_object(
            manager=fm,
            object_type=ObjectType.FUNCTION,
            object_definition=function,
            existing_objects=existing_functions,
            packages=packages,
            stage_artifact_path=artifact_stage_target,
        )
        deploy_status.append(operation_result)

    return CollectionResult(deploy_status)


def _find_existing_objects(
    object_type: ObjectType, objects: List[Dict], om: ObjectManager
):
    existing_objects = {}
    for object_definition in objects:
        identifier = build_udf_sproc_identifier(object_definition)
        try:
            current_state = om.describe(
                object_type=object_type.value.sf_name,
                name=remove_parameter_names(identifier),
            )
            existing_objects[identifier] = current_state
        except ProgrammingError:
            pass
    return existing_objects


def _check_if_all_defined_integrations_exists(
    om: ObjectManager, functions: List[Dict], procedures: List[Dict]
):
    existing_integrations = {
        i["name"].lower()
        for i in om.show(object_type="integration", cursor_class=DictCursor, like=None)
        if i["type"] == "EXTERNAL_ACCESS"
    }
    declared_integration: Set[str] = set()
    for object_definition in [*functions, *procedures]:
        external_access_integrations = {
            s.lower() for s in object_definition.get("external_access_integrations", [])
        }
        secrets = [s.lower() for s in object_definition.get("secrets", [])]

        if not external_access_integrations and secrets:
            raise SecretsWithoutExternalAccessIntegrationError(
                object_definition["name"]
            )

        declared_integration = declared_integration | external_access_integrations

    missing = declared_integration - existing_integrations
    if missing:
        raise ClickException(
            f"Following external access integration does not exists in Snowflake: {', '.join(missing)}"
        )


def _deploy_procedure_with_coverage(
    artifact_stage_directory: str,
    artifact_stage_target: str,
    build_artifact_path: Path,
    packages: List,
    pm: ProcedureManager,
    procedure: Dict,
    stage_manager: StageManager,
    stage_name: str,
    existing_objects: Dict[str, Dict],
):
    # This changes existing artifact, so we need to generate wrapper and only then deploy
    handler = _alter_procedure_artifact_with_coverage_wrapper(
        artifact_path=build_artifact_path,
        handler=procedure["handler"],
        identifier=build_udf_sproc_identifier(procedure),
        artifact_stage_path=artifact_stage_target,
        stage_name=stage_name,
    )
    stage_manager.put(
        local_path=build_artifact_path,
        stage_path=artifact_stage_directory,
        overwrite=True,
    )
    procedure["handler"] = handler
    packages.append("coverage")
    operation_result = _deploy_single_object(
        manager=pm,
        object_type=ObjectType.PROCEDURE,
        object_definition=procedure,
        packages=packages,
        stage_artifact_path=artifact_stage_target,
        existing_objects=existing_objects,
    )
    return CollectionResult([operation_result])


def get_app_stage_path(snowpark):
    artifact_stage_directory = (
        f"@{snowpark.get('stage_name', DEPLOYMENT_STAGE)}/{snowpark['project_name']}"
    )
    return artifact_stage_directory


def _deploy_single_object(
    manager: FunctionManager | ProcedureManager,
    object_type: ObjectType,
    object_definition: Dict,
    existing_objects: Dict[str, Dict],
    packages: List[str],
    stage_artifact_path: str,
):
    identifier = build_udf_sproc_identifier(object_definition)
    log.info(f"Deploying {object_type}: {identifier}")

    handler = object_definition["handler"]
    returns = object_definition["returns"]
    replace_object = False

    object_exists = identifier in existing_objects
    if object_exists:
        replace_object = check_if_replace_is_required(
            object_type,
            existing_objects[identifier],
            handler,
            returns,
        )

    if object_exists and not replace_object:
        return {
            "object": identifier,
            "type": str(object_type),
            "status": "packages updated",
        }

    create_or_replace_kwargs = {
        "identifier": identifier,
        "handler": handler,
        "return_type": returns,
        "artifact_file": stage_artifact_path,
        "packages": packages,
        "runtime": object_definition.get("runtime"),
        "external_access_integrations": object_definition.get(
            "external_access_integrations"
        ),
        "secrets": object_definition.get("secrets"),
    }
    if object_type == ObjectType.PROCEDURE:
        create_or_replace_kwargs["execute_as_caller"] = object_definition.get(
            "execute_as_caller"
        )

    manager.create_or_replace(**create_or_replace_kwargs)

    status = "created" if not object_exists else "definition updated"
    return {"object": identifier, "type": str(object_type), "status": status}


def _get_snowpark_artefact_path(snowpark_definition: Dict):
    source = Path(snowpark_definition["src"])
    artefact_file = Path.cwd() / (source.name + ".zip")
    return artefact_file


@app.command("build")
@global_options
@with_output
@with_project_definition("snowpark")
def build(
    pypi_download: str = PyPiDownloadOption,
    check_anaconda_for_pypi_deps: bool = CheckAnacondaForPyPiDependencies,
    package_native_libraries: str = PackageNativeLibrariesOption,
    **options,
) -> CommandResult:
    """
    Builds the Snowpark project as a `.zip` archive that can be used by `deploy` command.
    The archive is built using only `src` directory specified in project file.
    """
    snowpark = cli_context.project_definition
    source = Path(snowpark.get("src"))
    artefact_file = _get_snowpark_artefact_path(snowpark)
    log.info("Building package using sources from: %s", source.resolve())

    snowpark_package(
        source=source,
        artefact_file=artefact_file,
        pypi_download=pypi_download,  # type: ignore[arg-type]
        check_anaconda_for_pypi_deps=check_anaconda_for_pypi_deps,
        package_native_libraries=package_native_libraries,  # type: ignore[arg-type]
    )
    return MessageResult(f"Build done. Artefact path: {artefact_file}")


class _SnowparkObject(Enum):
    """This clas is used only for snowpark execute where choice is limited."""

    PROCEDURE = str(ObjectType.PROCEDURE)
    FUNCTION = str(ObjectType.FUNCTION)


def _execute_object_method(
    method_name: str,
    object_type: _SnowparkObject,
    **kwargs,
):
    if object_type == _SnowparkObject.PROCEDURE:
        manager = ProcedureManager()
    elif object_type == _SnowparkObject.FUNCTION:
        manager = FunctionManager()
    else:
        raise ClickException(f"Unknown object type: {object_type}")

    return getattr(manager, method_name)(**kwargs)


@app.command("execute")
@with_output
@global_options_with_connection
def execute(
    object_type: _SnowparkObject = ObjectTypeArgument,
    execution_identifier: str = execution_identifier_argument(
        "procedure/function", "hello(1, 'world')"
    ),
    **options,
) -> CommandResult:
    """Executes a procedure or function in a specified environment."""
    cursor = _execute_object_method(
        "execute", object_type=object_type, execution_identifier=execution_identifier
    )
    return SingleQueryResult(cursor)


def _replace_handler_in_zip(
    proc_name: str,
    proc_signature: str,
    handler: str,
    zip_file_path: str,
    coverage_reports_stage: str,
    coverage_reports_stage_path: str,
) -> str:
    """
    Given an existing zipped stored proc artifact, this function inserts a file containing a code coverage
    wrapper, then returns the name of the new handler that the proc should use
    """
    handler_module, _, handler_function = handler.rpartition(".")
    with TemporaryDirectory() as temp_dir:
        wrapper_file = os.path.join(temp_dir, "snowpark_coverage.py")
        snowcli.utils.file_utils.generate_snowpark_coverage_wrapper(
            target_file=wrapper_file,
            proc_name=proc_name,
            proc_signature=proc_signature,
            coverage_reports_stage_path=coverage_reports_stage_path,
            handler_module=handler_module,
            handler_function=handler_function,
        )
        add_file_to_existing_zip(zip_file=zip_file_path, file=wrapper_file)
    return "snowpark_coverage.measure_coverage"
