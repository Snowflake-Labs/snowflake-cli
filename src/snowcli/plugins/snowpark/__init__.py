from snowcli.plugins.snowpark.commands import app
from snowcli.plugins.snowpark.package.commands import app as package_app
from snowcli.plugins.snowpark.procedure_coverage.commands import (
    app as procedure_coverage_app,
)

app.add_typer(package_app)
app.add_typer(procedure_coverage_app)
