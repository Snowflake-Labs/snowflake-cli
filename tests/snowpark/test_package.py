import logging
import os
from unittest import mock
from unittest.mock import MagicMock, patch
from zipfile import ZipFile

import pytest
from snowflake.cli.plugins.snowpark.models import (
    Requirement,
    SplitRequirements,
)

from tests.test_data import test_data


class TestPackage:
    @pytest.mark.parametrize(
        "argument",
        ["snowflake-connector-python", "some-weird-package-we-dont-know"],
    )
    @patch("snowflake.cli.plugins.snowpark.package.anaconda.requests")
    def test_package_lookup(
        self, mock_requests, argument, monkeypatch, runner, snapshot
    ) -> None:
        mock_requests.get.return_value = self.mocked_anaconda_response(
            test_data.anaconda_response
        )

        result = runner.invoke(["snowpark", "package", "lookup", argument])

        assert result.exit_code == 0
        assert result.output == snapshot

    @patch("snowflake.cli.plugins.snowpark.package_utils.download_packages")
    @patch(
        "snowflake.cli.plugins.snowpark.package.anaconda.AnacondaChannel.parse_anaconda_packages"
    )
    def test_package_lookup_with_install_packages(
        self, mock_package, mock_download, runner, capfd, snapshot
    ) -> None:

        mock_package.return_value = SplitRequirements(
            [], [Requirement("some-other-package")]
        )
        mock_download.return_value = (
            True,
            SplitRequirements(
                [Requirement("snowflake-snowpark-python")],
                [Requirement("some-other-package")],
            ),
        )

        result = runner.invoke(["snowpark", "package", "lookup", "some-other-package"])
        assert result.exit_code == 0
        assert result.output == snapshot

    @patch("snowflake.cli.plugins.snowpark.package.commands.download_packages")
    @patch(
        "snowflake.cli.plugins.snowpark.package.commands.get_anaconda_from_snowflake"
    )
    @pytest.mark.parametrize(
        "extra_flags", [[], ["--skip-version-check"], ["--ignore-anaconda"]]
    )
    def test_package_create(
        self,
        mock_get_anaconda,
        mock_download,
        caplog,
        temp_dir,
        dot_packages_directory,
        runner,
        extra_flags,
    ) -> None:

        mock_anaconda = MagicMock(name="anaconda")
        mock_anaconda.is_package_available.return_value = False
        mock_get_anaconda.return_value = mock_anaconda
        mock_download.return_value = (
            True,
            SplitRequirements(
                [Requirement.parse("in-anaconda-package>=2")],
                [Requirement.parse("some-other-package")],
            ),
        )

        with caplog.at_level(
            logging.DEBUG, logger="snowflake.cli.plugins.snowpark.package"
        ):
            result = runner.invoke(
                ["snowpark", "package", "create", "totally-awesome-package"]
                + extra_flags
            )

        assert result.exit_code == 0, result.output
        assert "in-anaconda-package>=2" in result.output
        assert "some-other-package" not in result.output
        assert os.path.isfile("totally-awesome-package.zip"), result.output

        zip_file = ZipFile("totally-awesome-package.zip", "r")
        assert (
            "totally-awesome-package/totally-awesome-module.py" in zip_file.namelist()
        )
        os.remove("totally-awesome-package.zip")

        if "--ignore-anaconda" in extra_flags:
            expected_anaconda = expected_get_anaconda = []
        else:
            expected_get_anaconda = [mock.call()]
            expected_anaconda = [
                mock.call.is_package_available(
                    Requirement.parse("totally-awesome-package"),
                    skip_version_check="--skip-version-check" in extra_flags,
                )
            ]
        assert mock_anaconda.mock_calls == expected_anaconda
        assert mock_get_anaconda.mock_calls == expected_get_anaconda

    @mock.patch("snowflake.cli.plugins.snowpark.package.manager.StageManager")
    @mock.patch("snowflake.connector.connect")
    def test_package_upload(
        self,
        mock_connector,
        mock_stage_manager,
        package_file: str,
        runner,
        mock_ctx,
        mock_cursor,
    ) -> None:
        ctx = mock_ctx()
        mock_connector.return_value = ctx
        mock_stage_manager().put.return_value = mock_cursor(
            rows=[("", "", "", "", "", "", "UPLOADED")], columns=[]
        )

        result = runner.invoke(
            ["snowpark", "package", "upload", "-f", package_file, "-s", "stageName"]
        )

        assert result.exit_code == 0
        assert ctx.get_query() == ""

    @mock.patch(
        "snowflake.cli.plugins.snowpark.package.manager.StageManager._execute_query"
    )
    def test_package_upload_to_path(
        self,
        mock_execute_queries,
        package_file: str,
        runner,
        mock_ctx,
        mock_cursor,
    ) -> None:
        mock_execute_queries.return_value = MagicMock()

        result = runner.invoke(
            [
                "snowpark",
                "package",
                "upload",
                "-f",
                package_file,
                "-s",
                "db.schema.stage/path/to/file",
            ]
        )

        assert result.exit_code == 0
        assert mock_execute_queries.call_count == 2
        create, put = mock_execute_queries.call_args_list
        assert create.args[0] == "create stage if not exists db.schema.stage"
        assert "db.schema.stage/path/to/file" in put.args[0]

    @pytest.mark.parametrize(
        "flags",
        [
            ["--pypi-download"],
            ["-y"],
            ["--yes"],
            ["--pypi-download", "-y"],
        ],
    )
    @mock.patch(
        "snowflake.cli.plugins.snowpark.package.commands.get_anaconda_from_snowflake"
    )
    def test_lookup_install_flag_are_deprecated(self, _, flags, runner):
        result = runner.invoke(["snowpark", "package", "lookup", "foo", *flags])
        assert (
            "is deprecated. Lookup command no longer checks for package in PyPi"
            in result.output
        )

    @pytest.mark.parametrize(
        "flags",
        [
            ["--pypi-download"],
            ["-y"],
            ["--yes"],
            ["--pypi-download", "-y"],
        ],
    )
    @mock.patch(
        "snowflake.cli.plugins.snowpark.package.commands.get_anaconda_from_snowflake"
    )
    def test_create_install_flag_are_deprecated(self, _, flags, runner):
        result = runner.invoke(["snowpark", "package", "create", "foo", *flags])
        assert (
            "is deprecated. Create command always checks for package in PyPi."
            in result.output
        )

    @mock.patch(
        "snowflake.cli.plugins.snowpark.package.commands.get_anaconda_from_snowflake"
    )
    def test_lookup_install_with_out_flags_does_not_warn(self, _, runner):
        result = runner.invoke(["snowpark", "package", "lookup", "foo"])
        assert (
            "is deprecated. Lookup command no longer checks for package in PyPi"
            not in result.output
        )

    @staticmethod
    def mocked_anaconda_response(response: dict):
        mock_response = MagicMock()
        mock_response.status_code = 200
        mock_response.json.return_value = response

        return mock_response
