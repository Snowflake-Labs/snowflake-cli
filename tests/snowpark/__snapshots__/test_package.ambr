# serializer version: 1
# name: TestPackage.test_package_lookup[snowflake-connector-python]
  '''
  Package `snowflake-connector-python` is available in Anaconda. Latest available version: 2.7.12.
  
  '''
# ---
# name: TestPackage.test_package_lookup[some-weird-package-we-dont-know]
  '''
  
  Package `some-weird-package-we-dont-know` is not available in Anaconda. To prepare Snowpark compatible package run:
  snow snowpark package create some-weird-package-we-dont-know
  
  
  '''
# ---
